File: files/2026-05-17/CSEC 17052026.py
import pandas as pd

path = "2026-05-17/Challenge 121.xlsx"
input = pd.read_excel(path, usecols="B:C", nrows=13, skiprows=2)
test = pd.read_excel(path, usecols="E:F", nrows=3, skiprows=2)

input["Attendance"] = pd.to_datetime(input["Attendance"])

out = (
    pd.MultiIndex.from_product(
        [pd.date_range(input.Attendance.min(), input.Attendance.max(), freq="B"),
         input.Employee.unique()],
        names=["Date", "Employee"]
    )
    .to_frame(index=False)
    .merge(input.assign(k=1), left_on=["Date", "Employee"],
           right_on=["Attendance", "Employee"], how="left")
    .sort_values(["Employee", "Date"])
    .assign(g=lambda x: x.k.isna().groupby(x.Employee).cumsum())
    .dropna(subset="k")
    .groupby(["Employee", "g"])
    .size()
    .reset_index(name="gsize")
    .assign(gsize=lambda x: x.gsize.where(x.gsize > 1, 0))
    .groupby("Employee", as_index=False)["gsize"]
    .max()
    .rename(columns={"gsize": "Consecutive Streak"})
)
print(out["Consecutive Streak"].equals(test["Consecutive Streak (Work days)"]))
# True

File: files/2026-05-17/CSEC 17052026.R
library(tidyverse)
library(readxl)

path <- "2026-05-17/Challenge 121.xlsx"
input <- read_excel(path, range = "B3:C15")
test <- read_excel(path, range = "E3:F6")

df <- tibble(
    Date = seq.POSIXt(
        from = min(input$Attendance),
        to = max(input$Attendance),
        by = "day"
    )
) %>%
    filter(wday(Date) %in% c(2:6)) %>%
    crossing(Employee = unique(input$Employee)) %>%
    left_join(
        input %>% mutate(k = 1),
        by = c("Date" = "Attendance", "Employee")
    ) %>%
    arrange(Employee, Date) %>%
    mutate(Group = cumsum(is.na(k)), .by = Employee) %>%
    na.omit() %>%
    mutate(gsize = n(), .by = c(Employee, Group)) %>%
    summarize(
        `Consecutive Streak` = max(if_else(gsize > 1, gsize, 0L)),
        .by = Employee
    )
all.equal(df$`Consecutive Streak`, test$`Consecutive Streak (Work days)`)
# TRUE
